pkg/frontend/encryptionathost_validation.go

```diff
@@ -0,0 +1,72 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armfeatures"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/features"
+)
+
+type EncryptionAtHostValidator interface {
+	ValidateEncryptionAtHost(ctx context.Context, subscriptionID string, oc *api.OpenShiftCluster) error
+}
+
+type encryptionAtHostValidator struct{}
+
+func (e encryptionAtHostValidator) ValidateEncryptionAtHost(ctx context.Context, subscriptionID string, oc *api.OpenShiftCluster) error {
+	credential, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return err
+	}
+
+	subFeatureRegistrationsClient, err := armfeatures.NewSubscriptionFeatureRegistrationsClient(subscriptionID, credential, nil)
+	if err != nil {
+		return err
+	}
+	return validateEncryptionAtHost(ctx, subFeatureRegistrationsClient, oc)
+}
+
+func validateEncryptionAtHost(ctx context.Context, subFeatureRegistrationsClient features.SubscriptionFeatureRegistrationsClient, oc *api.OpenShiftCluster) error {
```

Comment:
(discussed during synchronous review)

Suggestion: This can be simplified as follows

```

func validateEncryptionAtHost(ctx context.Context, subFeatureRegistrationsClient features.SubscriptionFeatureRegistrationsClient, oc *api.OpenShiftCluster) error {
	if clusterRequiresEncryptionAtHost(oc.Properties.MasterProfile, oc.properties.WorkerProfiles) {
		return IsRegisteredForEncryptionAtHostFeature(ctx, subFeatureRegistrationsClient)
	}

	return nil
}

func clusterRequiresEncryptionAtHost(mp api.MasterProfile, wps []api.WorkerProfile) bool {
	if mp.EncryptionAtHost == api.EncryptionAtHostEnabled {
		return true
	}

	for _, wp := range workerProfiles {
		if wp.EncryptionAtHost == api.EncryptionAtHostEnabled {
			return true
		}
	}

	return false
}
```

Reply:
As discussed earlier, I have implemented an alternate logic. The suggested one looks great, but tried to achieve the same without having to implement another func here.